### PR TITLE
feat: add getCurrentPosition to component's ref

### PR DESCRIPTION
**Pull request by seyedmostafahasani**
## Summary
This function retrieves and returns the precise current position of the video playback, measured in seconds.

### Motivation
The goal is to provide users with the ability to access the current position of the video playback. This allows for the creation of custom controls like forward and rewind buttons, enhancing user experience.

### Changes
A new function `getCurrentPosition` has been added to the component's ref.

## Test plan

### CODE REVIEW

**Comment by freeboub** [comment]
Looks good to me, Just one note, I would prefer having the time in msec instead of sec. @KrzysztofMoch @YangJonghun what do you think of this point ?

File: android/src/main/java/com/brentvatne/react/VideoManagerModule.kt [hunk 1]
```diff
@@ -61,6 +62,13 @@ class VideoManagerModule(reactContext: ReactApplicationContext?) : ReactContextB
         }
     }
 
+    @ReactMethod
+    fun getCurrentPosition(reactTag: Int, promise: Promise) {
+        performOnPlayerView(reactTag) {
+           it?.getCurrentPosition(promise)
```
Comment (freeboub) [inline android/src/main/java/com/brentvatne/react/VideoManagerModule.kt]:
```suggestion
            it?.getCurrentPosition(promise)
```

**Comment by YangJonghun** [comment]
As far as I know, `onProgress` callback exists for the same reason. I wonder for what purpose this method was implemented.

**Comment by seyedmostafahasani** [comment]
@YangJonghun  you're correct! In situations where the user only needs to get the current position when interacting with custom controls, such as the rewind and forward buttons, it's not necessary to use a callback function that is called every progress update interval.

Instead, the user can use this approach that is only called when the user clicks the rewind or forward buttons. When a button is clicked, you can get the current position and then seek forward or backward based on the button that was clicked.

**Comment by YangJonghun** [comment]
@seyedmostafahasani 
That makes sense!
However, I think it might be a bit difficult to use it for precise work due to the limitations of old architecture, since the current time value keeps changing on playback.
It will be great later when we're done with the new architecture :)
Thanks for your explanation.👍

**Comment by seyedmostafahasani** [comment]
So now what is your opinion about the response of this function, second or millisecond?

**Comment by YangJonghun** [comment]
@seyedmostafahasani 
Oh sorry to late response.
I like second(specifically, second as a float value).
I prefer the same format as the value returned by `onProgress`, and I think many developers will be familiar with the web's video.currentTime format

**Comment by seyedmostafahasani** [comment]
I completely agree with you about second.

**Comment by freeboub** [comment]
As C developer I prefer working with integer, but if you are ok for sec, then OK for me !

File: android/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java [hunk 2]
```diff
@@ -685,6 +686,15 @@ private void initializePlayer() {
         mainHandler.postDelayed(mainRunnable, 1);
     }
 
+    public void getCurrentPosition(Promise promise) {
+        if (player != null) {
+            long currentPosition = player.getCurrentPosition() / 1000;
+            promise.resolve((double) currentPosition);
```
Comment (freeboub) [inline android/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: Directly use double here, if you cast to long, your will loose msec information !

```suggestion
            double currentPosition = player.getCurrentPosition() / 1000;
            promise.resolve(currentPosition);
```

**Review by freeboub — CHANGES_REQUESTED** [review]
See my comment for android

**Review by freeboub — APPROVED** [review]

File: docs/pages/component/methods.mdx [hunk 3]
```diff
@@ -100,6 +100,14 @@ tolerance is the max distance in milliseconds from the seconds position that's a
 
 This function will change the volume exactly like [volume](./props#volume) property. default value and range are the same then.
 
+### `getCurrentPosition`
+
+<PlatformsList types={['Android', 'iOS']} />
+
+`getCurrentPosition(): Promise<number>`
+
+This function retrieves and returns the precise current position of the video playback, measured in seconds.
```
Comment (freeboub) [inline docs/pages/component/methods.mdx:109]:
```suggestion
This function retrieves and returns the precise current position of the video playback, measured in seconds.
This function will throw an error if player is not initialized.
```

File: ios/Video/RCTVideo.swift [hunk 4]
```diff
@@ -1620,6 +1620,12 @@ class RCTVideo: UIView, RCTVideoPlayerViewControllerDelegate, RCTPlayerObserverH
         }
     }
 
+    @objc
+    func getCurrentPlaybackTime(_ resolve: @escaping RCTPromiseResolveBlock) {
+        let currentTime = RCTVideoUtils.getCurrentTime(playerItem: _playerItem)
+        resolve(currentTime)
```
Comment (freeboub) [inline ios/Video/RCTVideo.swift]: it would be great to match android behavior: if playerItem is nil, reject the promise.

**Review by freeboub — CHANGES_REQUESTED** [review]

**Review by freeboub — APPROVED** [review]

**Review by freeboub — APPROVED** [review]

**Comment by YangJonghun** [comment]
@freeboub @KrzysztofMoch @seyedmostafahasani 
I think we need to deprecate this feature
ref command with return value not supported by Fabric
(`save` command also have same problem)

**Comment by seyedmostafahasani** [comment]
@YangJonghun I think we can use them with Fabric, but an important point is that these are deprecated methods on the native or JS side. Utilizing deprecated methods on the native or JS side may result in these functions not working correctly. I recommend checking out `react-native-blurhash`, as it employs Fabric. This library uses functions that return values via promises, such as the `encode` function, which you can observe in these URLs.
https://github.com/mrousavy/react-native-blurhash/blob/master/src/index.tsx
https://github.com/mrousavy/react-native-blurhash/blob/master/android/src/main/java/com/mrousavy/blurhash/BlurhashModule.kt

**Comment by YangJonghun** [comment]
@seyedmostafahasani 
BlurHash is NativeModule.
TurboModule and Fabric are different. TurboModule can be return values
What I meant was that it's not possible to use a View as a ref and return it.

Currently we use `save` or `getCurrentPosition` in react-native-video is to reference a reactTag in NativeModule(VideoManager).
We are using `findNodeHandle` for above behavior, but findNodeHandle will be deprecated, so it is not recommended to continue using this method. We need to migrate to command method to replace it.
However command does not support return or callback functions.

**Comment by seyedmostafahasani** [comment]
@YangJonghun Oh yes, you're right. Thanks for the detailed explanation. I was referring to the deprecated method `findNodeHandle`. Could you please take a look at the `react-native-view-shot` library? I think this library could help us.

**Comment by YangJonghun** [comment]
@seyedmostafahasani 
Thanks for give me reference
But unfortunately react-native-view-shot also using `findNodeHandle` now. ([link](https://github.com/gre/react-native-view-shot/blob/master/src/index.js#L128))🥲

I guess I'll have to keep it for now and find a tricky way around it. Because we don't know when RN team get rid of findNodeHandle.
fyi) https://github.com/reactwg/react-native-new-architecture/blob/main/docs/enable-libraries-prerequisites.md

**Comment by seyedmostafahasani** [comment]
@YangJonghun anytime!
Did you see this [PR](https://github.com/gre/react-native-view-shot/pull/516)?

I agree with you, we can keep it for now. Also, if I find a better solution, I will share it with you.

**Pull request merged**